specification/dartLangSpec.tex

```diff
@@ -11677,113 +12041,209 @@ \subsection{Identifier Reference}
   \alt \STATIC{}
   \alt \TYPEDEF{}
 
-<IDENTIFIER\_START> ::= <IDENTIFIER\_START\_NO\_DOLLAR>
-  \alt `$'
+<IDENTIFIER\_START> ::= <IDENTIFIER\_START\_NO\_DOLLAR> | `$'
 
-<IDENTIFIER\_START\_NO\_DOLLAR> ::= <LETTER>
-  \alt `_'
+<IDENTIFIER\_START\_NO\_DOLLAR> ::= <LETTER> | `_'
 
 <IDENTIFIER\_PART\_NO\_DOLLAR> ::= \gnewline{}
-  <IDENTIFIER\_START\_NO\_DOLLAR>
-  \alt <DIGIT>
+  <IDENTIFIER\_START\_NO\_DOLLAR> | <DIGIT>
 
-<IDENTIFIER\_PART> ::= <IDENTIFIER\_START>
-  \alt <DIGIT>
+<IDENTIFIER\_PART> ::= <IDENTIFIER\_START> | <DIGIT>
 
 <qualified> ::= <identifier> (`.' <identifier>)?
 \end{grammar}
 
 \LMHash{}%
 A built-in identifier is one of the identifiers produced by the production
 \synt{BUILT\_IN\_IDENTIFIER}.
-It is a compile-time error if a built-in identifier is used as the declared name of a prefix, class, type parameter or type alias.
-It is a compile-time error to use a built-in identifier other than \DYNAMIC{} in a type annotation or type parameter.
+It is a compile-time error if a built-in identifier is used as
+the declared name of a prefix, class, type parameter or type alias.
+It is a compile-time error to use a built-in identifier
+other than \DYNAMIC{} or \FUNCTION{}
+in a type annotation or a type parameter bound.
```

Comment:
What does "*in* a type annotation or type parameter bound" mean?

Is `Function` "in" `List<notFunction> x;`?

Does this paragraph actually say anything? If you cannot declare a type named as a built-in identifier, then you can't use a built-in identifier as a type (unless it's `dynamic` or `Function`, which are types). So, what are we saying here that isn't already invalid?

Reply:
It used to be `as a type annotation or type parameter`, and it was adjusted to `in a type annotation..` because we wouldn't want to allow `List<static> x;`.

I added `as an identifier` to rule out the case `notFunction`.

Given that we'd want to have reasonably clear error messages, I think it's useful to say that we cannot use things like `List<static>` for syntactic reasons. If we do not say this then I'm sure you are right that it will turn out to be a compile-time error anyway, because we can't find a type that has the name `static`, but I still think it's going to be a less tool-team-friendly approach if we reach that result during static analysis (resolution), and require that the parser must survive this mal-terms first.

So I do think that it's useful to have a rule about these `<type>` cases, not just rules against declaration of types with those names.

Reply:
For spec simplicity, we can delete the sentence 12061-12063 and let tools implicitly rely on the fact that `List<static>` and such must be an error (and we don't specify whether they make the parser survive it and then fail to find a type, or they make it a syntax error).

Or we can keep it, because it documents that tools can rely on such things to be a compile-time error no matter what, which may simplify things for human readers of code as well as parsers.

@lrhn, what do you prefer?